Added login/signup/authentication backend API logic

f-hash opened the pull request:
Here is my new pull request

rnpnull commented on backend/trivia_api/static/signup.html:
```diff
@@ -0,0 +1,27 @@
+<form action="/signup" class="form">
```
get rid of this file, we shouldnt have any frontend functionality pushed on our side, only API implementation

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+    return jsonify({"code": 400, "message": "failed"})  # what am I returning here
+
+
+@login_manager.user_loader
```
this user loader is unnecessary, it's already in __init__, remove this

rnpnull commented on frontend/src/logins /login.js:
```diff
@@ -0,0 +1,11 @@
+import React from 'react'
```
get rid of this file + containing folder, isn't going to be used

rnpnull commented on backend/trivia_api/.gitignore:
```diff
@@ -0,0 +1 @@
+.env
```
this file is unnecessary? the .gitignore in the root folder has the same thing and it covers all subdirectory

rnpnull replied:
delete this file

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
     return dummy_users
 
 
+@blueprint.route("/signup", methods=["GET", "POST"])
```
```suggestion
@blueprint.route("/signup", methods=["POST"])
```

no need to allow get in the first place

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
 
+@blueprint.route("/signup", methods=["GET", "POST"])
+def signup_post():
+    if request.method == "POST":
```
```suggestion

```

take this if out after changing the decorator to only allow post

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+def signup_post():
+    if request.method == "POST":
+        # return jsonify({"message": "success"}, 200)
+        data = request.get_json("password")
```
```suggestion
        data = request.get_json()
```

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+                }
+            )
+    #  return jsonify({"new": newusers})
+    return jsonify({"code": 400, "message": "failed"})  # what am I returning here
```
```suggestion
    
```

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+        #  newusers2 = str(newusers)
+        newpasswords = data["password"]
+        passwords = generate_password_hash(newpasswords, method="sha256")
+        newuser = User(username=newusers)
+        newpass = User(password=passwords)
```
```suggestion
```

not sure what the goal was here,  but creating a User object requires a username and password, both strings. the way you pass these into the User later in the code is invalid, these dont need to be here

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+    if request.method == "POST":
+        # return jsonify({"message": "success"}, 200)
+        data = request.get_json("password")
+        newusers = data["username"]
+        #  newusers2 = str(newusers)
+        newpasswords = data["password"]
+        passwords = generate_password_hash(newpasswords, method="sha256")
```
```suggestion
        username = data["username"]
        password = data["password"]
```

variable naming, no need to hash if a user isnt going to be made

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+        passwords = generate_password_hash(newpasswords, method="sha256")
+        newuser = User(username=newusers)
+        newpass = User(password=passwords)
+        if (
+            User.query.filter_by(username=newuser.username).first() is None
+            and User.query.filter_by(password=newpass.password).first() is None
+        ):
```
```suggestion
        if not User.query.filter_by(username=username).first():
```

rewrite since the previous User objects are removed

also, you shouldnt check whether a user with the password exists; that would just disallow 2 people from having the same password which isnt what we want

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+            User.query.filter_by(username=newuser.username).first() is None
+            and User.query.filter_by(password=newpass.password).first() is None
+        ):
+            entry = User(username=newuser.username, password=newpass.password)
```
```suggestion
            new_user = User(username=username, password=generate_password_hash(password, method="sha256"))
```

variable naming/logic fix

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+            entry = User(username=newuser.username, password=newpass.password)
+            db.session.add(entry)
+            db.session.commit()
+            return jsonify({"code": 200, "message": "failed", "message": "success"})
```
no need to pass code, esp if 200, also the format I already have in place is to use success: True or if failure, False along with an error field
```suggestion
            return {"success": True}
```

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+            db.session.commit()
+            return jsonify({"code": 200, "message": "failed", "message": "success"})
+        else:
+            return jsonify(
+                {
+                    "code": 400,
+                    "message": "failed",
+                    "message": "User Name is already Taken",
+                }
+            )
```
```suggestion
            return {"success": False, "error": "username already taken" }
```

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+
+@blueprint.route("/login", methods=["GET", "POST"])
+def login_post():
+    username = request.get_json("username")
```
```suggestion
    username = request.get_json()
```

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+@blueprint.route("/login", methods=["GET", "POST"])
+def login_post():
+    username = request.get_json("username")
+    checkusers = username["username"]
+    password = username["password"]
```
```suggestion
    username = data["username"]
    password = data["password"]
```

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+    checkusers = username["username"]
+    password = username["password"]
+    #     remember = bool(request.form.get("remember"))
+    user = User.query.filter_by(username=checkusers).first()
```
```suggestion
    user = User.query.filter_by(username=username).first()
```

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+                "message": "You have entered the wrong login! Try again",
+            }
+        )
+    #         flash("Please check your login details and try again.")
+    #         return redirect(url_for("blueprint.login"))
```
```suggestion
```

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+    #     remember = bool(request.form.get("remember"))
+    user = User.query.filter_by(username=checkusers).first()
+    if not user or not check_password_hash(user.password, password):
+        return jsonify(
+            {
+                "code": 400,
+                "message": "failed",
+                "message": "You have entered the wrong login! Try again",
+            }
+        )
```
```suggestion
        return {"success": False, "error": "invalid login"}
```

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+        )
+    #         flash("Please check your login details and try again.")
+    #         return redirect(url_for("blueprint.login"))
+    else:
+        login_user(user)
+        return jsonify({"code": 200, "message": "failed", "message": "success"})
```
```suggestion
    login_user(user)
    return {"success": True}
```

else is unnecessary because of return from the if

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -61,6 +61,71 @@ def users(user_id):
+@login_required
+def logout():
+    logout_user()
+    return jsonify({"message": "Success"})
```
```suggestion
    return {"success": True}
```

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -55,88 +55,141 @@ def users(user_id):
-#     if not user or not check_password_hash(user.password, password):
-#         flash("Please check your login details and try again.")
-#         return redirect(url_for("blueprint.login"))
+@blueprint.route("/login", methods=["GET", "POST"])
```
```suggestion
@blueprint.route("/login", methods=["POST"])
```

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -55,88 +55,141 @@ def users(user_id):
-#     username = request.form.get("username")
-#     password = request.form.get("password")
-#     remember = bool(request.form.get("remember"))
+def loaduser(user_id):
+    return User.query.get(user_id)
```
u gotta remove this whole method not just the decorator

rnpnull commented on backend/trivia_api/views.py:
```diff
@@ -55,88 +55,141 @@ def users(user_id):
-#     login_user(user, remember=remember)
-#     return redirect(url_for("blueprint.movies"))
 
+@blueprint.route("/logout", methods=["GET", "POST"])
```
just needs to be

`@blueprint.route("/logout")`

f-hash commented:
Changed the login/signup methods by rewriting the database logic and by making rewriting the signup logic to only accept acceptable JSON requests, and further deleted the load user methods and changed the logout method, In addition, fixing the merge conflicts between login/signup and main
